// @dcloudio/uni-id/src/lib/common/login-exec.js
import {
  userCollection
} from '../utils/config'

export default async function (user, options = {}) {
  if (user.status === 1) {
    return {
      code: 10001,
      msg: '账号已禁用'
    }
  }

  const config = this._getConfig()
  // 过期token清理
  let tokenList = user.token || []
  // 兼容旧版逻辑
  if (typeof tokenList === 'string') {
    tokenList = [tokenList]
  }
  const expiredToken = this._getExpiredToken(tokenList)
  tokenList = tokenList.filter(item => {
    return expiredToken.indexOf(item) === -1
  })

  let tokenInfo
  if (config.removePermissionAndRoleFromToken) {
    const needPermission = options.needPermission
    tokenInfo = this.createToken({
      uid: user._id,
      needPermission
    })
  } else {
    const role = user.role || []
    let permission
    if (role.length === 0 || role.includes('admin')) {
      permission = []
    } else {
      permission = await this._getPermissionListByRoleList(role)
    }
    tokenInfo = this.createToken({
      uid: user._id,
      role,
      permission
    })
  }

  const {
    token,
    tokenExpired
  } = tokenInfo

  tokenList.push(token)
  user.token = tokenList

  const updateData = {
    last_login_date: Date.now(),
    last_login_ip: __ctx__.CLIENTIP,
    token: tokenList,
    ...options.extraData
  }
  await userCollection.doc(user._id).update(updateData)

  const userInfo = Object.assign({}, user, updateData)

  return {
    code: 0,
    msg: '登录成功',
    token,
    uid: userInfo._id,
    username: userInfo.username,
    type: 'login',
    userInfo,
    tokenExpired
  }
}
