// @dcloudio/uni-id/src/lib/login/login-by-alipay.js
import {
  userCollection
} from '../utils/config'

async function loginByAlipay (code) {
  let params = code
  if (typeof code === 'string') {
    params = {
      code
    }
  }
  const needPermission = params.needPermission
  const clientPlatform = params.platform || __ctx__.PLATFORM
  const {
    openid
  } = await this._getAlipayApi({
    platform: clientPlatform
  }).code2Session(params.code)
  if (!openid) {
    return {
      code: 10501,
      msg: '获取openid失败'
    }
  }
  const userList = await userCollection.where({
    ali_openid: openid
  }).get()
  // openid已注册
  if (userList && userList.data && userList.data.length > 0) {
    const userMatched = userList.data[0]

    const loginExecRes = await this._loginExec(userMatched, {
      needPermission
    })
    if (loginExecRes.code !== 0) {
      return loginExecRes
    }
    const {
      userInfo
    } = loginExecRes

    return {
      ...loginExecRes,
      openid,
      mobileConfirmed: userInfo.mobile_confirmed === 1,
      emailConfirmed: userInfo.email_confirmed === 1
    }
  } else {
    const user = {
      ali_openid: openid
    }
    user.my_invite_code = params.myInviteCode
    user.role = params.role
    const registerExecResult = await this._registerExec(user, {
      needPermission
    })
    if (registerExecResult.code !== 0) {
      return registerExecResult
    }
    return {
      ...registerExecResult,
      openid,
      mobileConfirmed: false,
      emailConfirmed: false
    }
  }
}

export default loginByAlipay
